File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/MaxSubArray.java
package com.only4.algorithm.version4j.leetcode;

/**
 * @author zhenyu.jiang
 */
public class MaxSubArray {
    public int maxSubArray(int[] nums) {
        int currentSum = nums[0];
        int maxSum = nums[0];

        for (int i = 1; i < nums.length; i++) {
            // 状态转移：选择加入前面的子数组或重新开始
            currentSum = Math.max(nums[i], currentSum + nums[i]);
            // 更新全局最大和
            maxSum = Math.max(maxSum, currentSum);
        }

        return maxSum;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/maxSubArray.kt
package com.only4.algorithm.version4k.leetcode

fun maxSubArray(nums: IntArray): Int {
    var currentSum = nums[0]
    var maxSum = nums[0]

    for (i in 1 until nums.size) {
        // 状态转移：选择加入前面的子数组或重新开始
        currentSum = maxOf(nums[i], currentSum + nums[i])
        // 更新全局最大和
        maxSum = maxOf(maxSum, currentSum)
    }

    return maxSum
}
